qiskit/algorithms/minimum_eigensolvers/adapt_vqe.py

```diff
@@ -0,0 +1,311 @@
+    """A class enumerating the various finishing criteria."""
+
+    CONVERGED = "Threshold converged"
+    CYCLICITY = "Aborted due to a cyclic selection of evolution operators"
+    MAXIMUM = "Maximum number of iterations reached"
+
+
+class AdaptVQE(VariationalAlgorithm):
+    """The Adaptive Variational Quantum Eigensolver algorithm.
+
+    `AdaptVQE <https://arxiv.org/abs/1812.11173>`__ is a quantum algorithm which creates a compact
+    ansatz from a set of evolution operators. It iteratively extends the ansatz circuit, by
+    selecting the building block that leads to the largest gradient from a set of candidates. In
+    chemistry, this is usually a list of orbital excitations. This results in a wavefunction ansatz
+    which is uniquely adapted to the operator whose minimum eigenvalue is being determined.
+    This class relies on a supplied instance of :class:`~.VQE` to find the minimum eigenvalue.
+    The performance of AdaptVQE significantly depends on the minimization routine.
+
+    .. code-block:: python
+
+      from qiskit.algorithms.minimum_eigensolvers import AdaptVQE, VQE
+      from qiskit.algorithms.optimizers import SLSQP
+      from qiskit.primitives import Estimator
+      from qiskit.circuit.library import EvolvedOperatorAnsatz
+
+      # get your Hamiltonian
+      hamiltonian = ...
+
+      # construct your ansatz
+      ansatz = EvolvedOperatorAnsatz(...)
+
+      vqe = VQE(Estimator(), ansatz, SLSQP())
+
+      adapt_vqe = AdaptVQE(vqe)
+
+      eigenvalue, _ = adapt_vqe.compute_minimum_eigenvalue(hamiltonian)
+
+
+    Attributes:
+        solver: a :class:`~.VQE` instance used internally to compute the minimum eigenvalues.
+            It is a requirement that the :attr:`~.VQE.ansatz` of this solver is of type
+            :class:`qiskit.circuit.library.EvolvedOperatorAnsatz`.
+        threshold: the convergence threshold for the algorithm. Once all gradients have an absolute
+            value smaller than this threshold, the algorithm terminates.
+        max_iterations: the maximum number of iterations for the adaptive loop. If ``None``, the
+            algorithm is not bound in its number of iterations.
+    """
+
+    def __init__(
+        self,
+        solver: VQE,
+        *,
+        threshold: float = 1e-5,
+        max_iterations: int | None = None,
+    ) -> None:
+        """
+        Args:
+            solver: a :class:`~.VQE` instance used internally to compute the minimum eigenvalues.
+                It is a requirement that the :attr:`~.VQE.ansatz` of this solver is of type
+                :class:`qiskit.circuit.library.EvolvedOperatorAnsatz`.
+            threshold: the convergence threshold for the algorithm. Once all gradients have an
+                absolute value smaller than this threshold, the algorithm terminates.
+            max_iterations: the maximum number of iterations for the adaptive loop. If ``None``, the
+                algorithm is not bound in its number of iterations.
```

Comment:
So the stopping criteria is threshold but maybe terminated sooner if max_iterations is reached. If the latter is None can this go on indefinitely if it never reaches threshold value i.e. is there any safety net or entirely down to the user?

I note the scipy optimizers use `tol` for termination tolerance, but maybe its better this is different to avoid any confusion with tol on the optimizer used by the solver.

Reply:
Yes, the algorithm can potentially run forever. Whether or not it will run into `cyclicity` at that point is hard to say. It may still take a **very** long time.

We can also set a default value for `max_iterations` other than `None`, if you want to go with a "safer" option.

Reply:
Lets leave it be for now as I imagine a non-None default would take time. Maybe at some point we could do like the optimizers do and add a callback at each iteration where passing back True from this would stop the optimization. This way it can be externally monitored/stopped as well. This would be some future improvement/feature should we chose to do it, not now though.